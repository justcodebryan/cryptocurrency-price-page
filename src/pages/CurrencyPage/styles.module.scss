.currency {
  &-title {
    margin: 30px 0 24px 24px;
    word-wrap: break-word;
  }

  &-list {
    width: calc(100% - 12px * 2);
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(320px, 1fr));
    grid-gap: 48px;
    grid-auto-rows: minmax(160px, auto);

    @media screen and (max-width: (400px)) {
      display: flex;
      flex-direction: column;
      width: 100%;
    }

    &-item {
      padding: 12px 0 8px 16px;
      box-sizing: border-box;

      &-title {
        word-wrap: break-word;
      }

      &-price {
        width: calc(100% - 8px);
        line-height: 1.5;
        color: rgb(245 165 30);
        white-space: nowrap;
        text-overflow: ellipsis;
        -o-text-overflow: ellipsis;
        overflow: hidden;
      }

      &-volume {
        width: calc(100% - 8px);
        color: rgb(150 150 150);

        & > p {
          width: 100%;
          line-height: 1.5;
          font-weight: 600;
        }

        & > div {
          white-space: nowrap;
          text-overflow: ellipsis;
          -o-text-overflow: ellipsis;
          overflow: hidden;
        }
      }
    }
  }
}
