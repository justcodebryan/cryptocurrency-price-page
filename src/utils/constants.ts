import { NoopFn } from '@/types'

export const NOOP: NoopFn = () => void 0

export const DEFAULT_MODAL_CONTAINER_ID = 'modal-container'

export const DEFAULT_POLLING_INTERVAL = 10 * 1000
