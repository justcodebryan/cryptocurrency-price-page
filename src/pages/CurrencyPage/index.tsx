import Card from '@/components/Card'
import Section from '@/components/Section'
import Title from '@/components/Title'
import Typography from '@/components/Typography'
import useEffectOnce from '@/hooks/useEffectOnce'
import { fetchCurrencyList } from '@/services/currency'
import { Currency, CurrencyQueryString } from '@/types/currency'
import { DEFAULT_POLLING_INTERVAL } from '@/utils/constants'
import { useState } from 'react'

import styles from './styles.module.scss'

const CurrencyPage = () => {
  const [currencyList, setCurrencyList] = useState<Currency[]>([])

  useEffectOnce(() => {
    const getCurrencyList = async () => {
      const res = await fetchCurrencyList(CurrencyQueryString)
      setCurrencyList(res)
    }

    const interval = setInterval(() => {
      getCurrencyList()
    }, DEFAULT_POLLING_INTERVAL)

    getCurrencyList()

    return () => {
      clearInterval(interval)
    }
  })

  return (
    <>
      <Title level={1} className={styles['currency-title']}>
        Cryptocurrency Realtime Price
      </Title>
      <Section className={styles['currency-list']}>
        {(currencyList || []).map((currency) => (
          <Card key={currency.asset_id}>
            <div className={styles['currency-list-item']}>
              <Title level={1} className={styles['currency-list-item-title']}>
                {currency.name}
              </Title>
              <Title level={2} className={styles['currency-list-item-price']}>
                ${currency.price_usd}
              </Title>
              <Typography className={styles['currency-list-item-volume']}>
                <p>volume:</p>
                {currency.volume_1day_usd}
              </Typography>
            </div>
          </Card>
        ))}
      </Section>
    </>
  )
}

export default CurrencyPage
