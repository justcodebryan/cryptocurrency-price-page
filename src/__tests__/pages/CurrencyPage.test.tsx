import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import CurrencyPage from '@/pages/CurrencyPage'
import { toJSON } from '@/utils/test-utils'

describe('Currency Page Test Suit', () => {
  test('Should show title', () => {
    render(<CurrencyPage />)
    expect(screen.getByRole('heading', { name: 'Cryptocurrency Realtime Price', level: 1 })).toBeDefined()
  })

  test('Should show list but do not have item', async () => {
    const component = renderer.create(<CurrencyPage />)
    const componentInstance = component.root
    expect(componentInstance.children).toHaveLength(1)
  })

  test('Should not show the content at the start', () => {
    expect(screen.getByText(/BTC/i)).toBeUndefined()
  })

  test('Currency Page Snapshot', () => {
    const component = renderer.create(<CurrencyPage />)
    const tree = toJSON(component)
    expect(tree).toMatchSnapshot()
  })
})
