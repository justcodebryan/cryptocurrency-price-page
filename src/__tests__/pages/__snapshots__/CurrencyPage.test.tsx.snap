// Vitest Snapshot v1

exports[`Currency Page DOM Test 1`] = `
<div
  className="_box_6c0536"
>
  <h1
    className="_title_35ba99 _title-1_35ba99 _currency-title_f561ab"
  >
    Cryptocurrency Realtime Price
  </h1>
  <div
    className="_section_8f1029 _currency-list_f561ab"
  />
</div>
`;

exports[`Currency Page Test Suit > Currency Page Snapshot 1`] = `
<div
  className="_box_6c0536"
>
  <h1
    className="_title_35ba99 _title-1_35ba99 _currency-title_f561ab"
  >
    Cryptocurrency Realtime Price
  </h1>
  <div
    className="_section_8f1029 _currency-list_f561ab"
  />
</div>
`;
